### ISSUE TITLE
'Compact session' and 'Redraw screen' as external commands

### ISSUE THREAD

**Issue by marttt** [original-male/non#226]
As explained in #157, I use Non for editing radio programs.
Typically this involves heavy cut-and-paste,
so a 'ripple editing' feature is immensely useful.

I made a poor man's version of this in Awk:
http://tihnik.sdf.org/s/non/non-ripple

That works well for my needs, but it raised a thought:
would it be possible to have the 'compact session'
and 'redraw screen' commands available externally?

Thanks to Non's wonderful session format,
this would give a nice primitive scripting mechanism:

    compact-session | modify-compacted-session-with-awk-sed-etc | reload-that-same-session-in-timeline

I've written a few simple hobbyst-level scripts (http://tihnik.sdf.org/s/non/), but currently I only see my modifications after closing Timeline and then re-opening it by hand.

Or is there already some way to call the 'compact session' command as an external process, from outside a Timeline window, from within a shell script?

**Comment by original-male** [comment]
There isn't. However, every time NT terminate normally (i.e. `SIGTERM`, not `SIGKILL` or a segfault), it saves a compacted version of the history for faster project loading. This is called `snapshot`. The compaction operation is basically just creating the snapshot and replacing the `history` with it. So, if you had a script that started non-timeline, it could detect that it had exited, fiddle with the snapshot, then replace the history with the snapshot, and finally relaunch a new instance of non-timeline. 

Of course, I couldn't recommend working like this, as you loose all of the advantages of journaling. What you really "should" be doing is only appending the state changes you want to make to the history file and deleting the (now invalidated) snapshot.

As far as a dynamic scripting interface goes, getting that right would not be trivial. Typically this involves pulling in some scripting language dependency like Lua... Most likely it would be just another way to shoot oneself in the foot.

I have toyed with the idea of adding a range-based delete and insert time option (affecting all tracks), but that doesn't seem to be exactly what you're trying to do. A function in NT to abut all regions on a track together would be pretty trivial to implement, and would be journaled and undo-able. Is that the only operation like this you commonly need to perform?

**Comment by original-male** [comment]
@marttt I took a look at your scripts: You're doing some cool stuff. I do like the concept, and yes, I did design the format to be easily sed/awk-able and as human readable as is possible given the requirements. 

I'm thinking that one possibility would be to have an option to launch scripts as filters from within NT. NT could pass a session snapshot in on `stdin` and read from the script's `stdout`, then apply the changes to the current state, journaling them appropriately. Obviously there would still be potential for corruption if you tried to e.g. create new objects with IDs that are already in use, but I think such as system could cover the functionality of many of your scripts. The ones that manipulate the audio sources themselves would be problematic. For one thing, a script should never change an audio source in place... Everything must be immutable, with only the creation of new copies with new names being allowed...

**Comment by marttt** [comment]
@original-male Thanks!

'Scripts as filters from within NT' sounds excellent.
This would encourage the user to hand-tailor basic things with one-liners,
think 'the Unix way',
whilst also keeping the software lean.

As for my specific needs -- well, I tried to accomplish something like this:
https://www.youtube.com/watch?v=DS3GF41QUxs

Radio editing can involve heavy re-organization of source material.
Cut a bunch of clips here to paste them over there instead.
After this, hide the cutting points from the listener as well as you can.
The result has to be a seamless stream of audio.

With ripple editing, there's no need to 'glue' stuff back together by hand.
I cut something out, and other clips shift right next to one another
by themselves.
This way editing can be rapid fast.

Then again, I hesitated asking for this feature from you, because:
1. I thought heavy 'rippling' would make the journal file grow HUGE
2. No need to bloat the software; maybe it would also slow NT down
significantly?
2. Doing a poor man's version seemed trivial and made me learn some more Awk :)
3. Now I sort of like the way my current solution slows me down during editing.

I suppose every added feature should be as universal as possible.
That said, 'range-based insert and delete time' seems worth considering as well.
So, e.g, I could command Non to 'insert 2 minutes of silence (=time)
at cursor', right?

This would, indeed, provide a robust solution for my essential needs:
a. Select a bunch of clips,
b. thus roughly, by eye, evaluate their total length,
c. then insert same amount of silence elsewhere,
d. and then drag the selected clips over to that 'silent' area,
e. and later 'glue' clips together with my non-ripple.awk script

The advantage of this (over a dedicated 'ripple editing' function)
seems to be greater universality. More user cases, I guess.
I suppose this 'add silence' feature would come in handy for musicians as well.

And, I know next to nothing about serious coding,
but 'add time' is probably also much easier to implement than rippling,
which is essentially an advanced cut-and-copy-and-paste, isn't it?

(I suppose there is a reason for why Non doesn't offer a 'cut'
function thus far?)

**Comment by original-male** [comment]
Regarding the history file size... That's not really a concern, the history file is always going to be far smaller than your smallest audio source, and the length of the history is already optimized away by the snapshot mechanism. It's only after a crash that the full history must be replayed... I have never found it necessary to compact a real session.

Regarding the insert/delete time function: It's not exactly trivial to implement because there's no guarantee that the selected range start/stop point will fall directly between region edges, and where it doesn't, some decision must be made (split it? extend a looped region? etc.). The place this most commonly comes up in music is at the beginning of a song, and that case is already covered by leaving some lead-in space when you start recording and setting the adjusting the export start point later... Which is why I've felt a pressing need to have insert/delete time. However, there have been a few times when it would have come in handy and I had to manually split a bunch of regions and move them.

Regarding cut & paste... That's kind of a philosophical issue. I personally despise the whole "desktop" metaphor, including the idea of a "clipboard". In real life a clipboard is a physical object which you can see... In the computer metaphor, it's an invisible place from which things tend to accidentally vanish. So I prefer to have mark, move, delete and duplicate operations... Which allow you to accomplish all the same things, but deleting objects is always explicit (not implicit where you move them to a clipboard and then forget about them.) It also allows for moves to easily be represented as such in the history (not as the object disappearing in one place and later re-appearing in another.)

Anyway, a "ripple" function which just (as a one time operation) abuts all regions on a track would be pretty trivial to implement. Easier than a scripting interface or an insert/delete time function.

**Comment by marttt** [comment]
I definitely resonate with the 'no clipboard' way of thinking;
thanks for the explanation.

As for the 'ripple' function: I would surely use it heavily.
Big thanks from me if you find the time to add this.
That, along with leaving some lead-in space just like you described,
would actually cover most of my editing needs.

I actually planned to write two sed helpers with the latter:
'non-append' and 'non-prepend'.
These should add empty space/some marked clips/etc to the beginning or
end of a session.

As a side note, I imagine the ":color" tag would allow rather
interesting algorithmic compositions:
re-color some clips in NT -> mess around with the re-colored spots
using unix tools.

So, yeah, awk-ing or sed-ing NT sessions is a great way to procrastinate. :)
A great session format you have, indeed!

cheers,
Mart

**Comment by original-male** [comment]
@marttt 

Yeah, I thought your use of the color property was very clever. I usually operate in region-color-inherits-track-color mode, where track color roughly encodes the subject matter. I.e. instrument type, so I probably wouldn't have thought of using region color as a kind of selection set.

I'd be interested in learning more about how you use Non in your production. If you'd like to write an article about it, I could put it up on the wiki.

Do you feel that there's any advantage to having that instantaneous/interactive ripple mode vs doing the abutment as a discrete operation? Another thing I'm curious about: when you're using NT in this way, are you working with multiple tracks, or just one track of dialogue? I've never attempted to edit this kind of content.

I was thinking that the most sensible/versatile way to implement it would be to have an "distribute selected regions" operation, which would set the space between the selected regions to a specified interval. I.e. 0, would abut them, -50 (frames?) would make them overlap by the given amount (useful if you planned to cross-fade the transitions), +50 would distribute with the given amount of space between regions.

**Comment by marttt** [comment]
@original-male Re: region coloring: thanks!
I've worked with several DAWs,
and somehow I've always preferred colors over markers to highlight
relevant parts of the montage.
So as soon as I realized the simplicity of NT's session format,
playing with this was the first and most natural thing to do.

E.g. another of my awk scripts, 'non-boring',
calculates the full length of all brown-colored regions.
Brown has traditionally been my color to mark parts of the source audio
that I am willing to throw out to make the program shorter.
Feeding that script the 'history' or 'snapshot' file
and piping the result through a 'samples2seconds' one-liner,
I can easily estimate how much not-so-relevant material I currently have.
I didn't try to do this in Pro Tools, Reaper or Ardour,
though I felt the need occasionally.
(Need for an additional feature, that, is, yes-yes. :)
But using NT, all I needed was an an one-liner in awk.
Dat awk which is available everywhere.
Just beautiful!

I think playing with the ':selected' parameter
might also open interesting possibilities regarding 'composing with
awk-powered NT' :),
etc, etc.

I've no CS background; programming-wise I'm just a hobbyst,
and awk is still pretty much my first language.
But I do feel that NT is wonderfully flexible for the things I do.
And, having just read E. S. Raymond's "The Art of Unix Programming",
I'd say it looks like a great example of unix-y software,
considering both the UI and the modular design logic.
It encourages simple, creative, modular thinking from any end-user actually.

So, sure, I'd be delighted to share my pathetic hacks for the wiki.
Maybe these rudimentary scripts would then inspire people with better
coding skills to go further.

Re: interactive ripple:
This is a nice discovery, but I've actually come to prefer the
discrete workflow.
Interactivity would add speed.
Editing would be rapid fast; you could "think with the mouse".
I suppose people working in a newsroom might prefer this.
But for the kind of work I mostly do
-- long, in-depth shows that should also create an "atmosphere" for
the listener --
I realized whilst using NT that it's actually better to be slower.
Interactive mode would encourage fiddling a lot with the content.
Shifting selections here, then anxiously shifting them back there
because you didn't like it, etc.
Discrete abutment, in my view, encourages a 'think first, then think
again, then maybe do' approach.
Fiddle less, hypothize and think more.
I'm a "slow person" in general,
so all in all, this tends to result in better work.
And the process is more enjoyable, kind-of zen-like, even.

Another thing I like about NT in that regard
is that it has so little keyboard/mouse shortcuts.
I'm a recovering vim and emacs user,
but I've much come to prefer a 'one well-thought-out key for one thing' approach
over recent years.

I think it relates to the 'Keyboard vs mouse' discussion --
does mousing free mental focus compared to keyboarding? --
which I'm interested in because I find the Plan 9 OS compelling:
http://plan9.bell-labs.com/wiki/plan9/Mouse_vs._Keyboard/index.html

**"The basic summary is cursoring around required a higher level of
mental planning to organize the interaction, which apparently obscures
the perception of the passage of time--think of being deeply engaged
in something and being surprised when you look at a clock-- whereas
the use of the mouse was done at a lower, mechanical level that left
the mind free for higher things, such as complaining about the
mouse."**

Re: "distribute selected regions" command:
This sounds great, very flexible.
Particularly if it operates on selected regions,
not the entire track.
That's because another thing I need to do
is to trim only a few samples at region boundaries.
(To make the listening experience smooth without using fades;
for some reason my Senior Producer doesn't like fades.)
I was thinking of writing a small CLI tool for this some day,
and I probably will, even if it's just an exercise for fun.
But I think specifiyng the distribution inverval
might help with that kind of microscopic trimming as well,
thus actually eliminating the need for a special tool.

Would the interval variable be freely entered by the user,
or would you give a dropdown menu with a bunch of intervals to choose from?
(Just like zooming currently works in NT?)

**Comment by original-male** [comment]
@marttt 

It was actually my frustrations with the Ardour session format (XML hell that kept getting corrupted because the whole structure is held in memory and rewritten upon save.) That inspired me to develop my own DAW in the first place...

I'm not sure about the interval. My instinct would be to use beats, but I doubt that unit has any relevance in the context of editing dialogue. So maybe samples or milliseconds. I was just thinking it would be a little popup number entry, like the one for edit time/tempo markers, preserving the previously entered value. So the whole sequence of operations would be: select some regions, execute "distribute selected regions" command (context menu or keystroke), enter new amount into popup or just click OK/press enter."

Regarding trim vs. fade... I think this is another situation where the needs are different for music. Usually when making music one is snapping regions to beats/bars, and it's therefore often preferable to have a region of the desired length beginning at the bar/beat line and faded in, than to have one trimmed perfectly and placed off of the bar/beat lines. So I'm accustomed to relying heavily on fades. Don't forget that there are different fade curves to choose from which can provide a more abrupt transition than the default.

**Comment by marttt** [comment]
@original-male
Regarding Ardour -- same here, and I think this experience is shared by many.
Ardour has ripple editing, but in practice it was crashing constantly
for me, and eating way too much RAM (I like to work with old, low-end
Thinkpads).

It was impossible to edit a 50-minute show with Ardour 2 or 3; no idea
how things might have developed by now.

I first switched to Reaper -- which is great and amazingly compact,
but required Wine back then. (They should have a pure Linux beta by
now as well.) When I saw Non, I was instantly hooked by the
modularity, and the simplicity of its GUI.

Re: intervals:
For dialogue, I think samples would be the most universal unit.
The popup would be fine, although it occurred to me that a dropdown
with a few fixed values (one of these being 0) and keybindings would
allow very quick editing.

This would be similar to the way zooming currently works:
one can press 1, 2 or 3. I use these quite a lot.

Maybe "0" could be the only value with a special key?

Then again, shortcut keys are by no means crucial;
adding these is surely a slippery slope.
Just a popup would be fine as well, and definitely more universal and
flexible for different uses.

**Comment by marttt** [comment]
Re: interval unit:
maybe this should be bound to the unit used on one of the two counters?
like a "current measuring unit".
